## I. Tự Luận

### `F1 Score`

In [1]:
from F1_score import *

In [2]:
print(f"Precision is : {precision(2, 3): .2f}")
print(f"Recall is : {recall(2, 4): .2f}")
print(f"F1 score is : {f1_score(2, 3, 4): .2f}")

Precision is :  0.40
Recall is :  0.33
F1 score is :  0.36


In [3]:
exercise1 = f1_score('a', 3, 4)
exercise1 = f1_score(2, 'a', 4)
exercise1 = f1_score(2, 3, 'a')

exercise1 = f1_score(2, 3, 0)

tp must be an integer
fp must be an integer
fn must be an integer
tp and fp and fn must be greater than 0


### `Activation`

In [50]:
from Activation import *

In [5]:
print(f"Activation function: {is_number(1.5)}")

Activation function: 0.8175744761936437


In [6]:
print(f"Activation function: {is_number('a')}")

x must be a number
Activation function: False


In [7]:
print(f"Activation function: {is_number(1)}")

Activation function: 0.7310585786300049


### `Regression Loss`

In [3]:
from Regression_loss import *

In [2]:
num_samples = input("Enter the number of samples: ")
regression_loss_func(numb_samples=num_samples)

Loss name: MAE
Sample: 0 | Prediction: 7.3646815087506345 | Target: 8.072575600059068 | Loss value: 0.7078940913084333

Loss name: MAE
Sample: 1 | Prediction: 6.513022726458624 | Target: 6.045157123269751 | Loss value: 0.4678656031888728

Loss name: MAE
Sample: 2 | Prediction: 3.8148993584079474 | Target: 0.2914970066645928 | Loss value: 3.5234023517433544

Loss name: MAE
Sample: 3 | Prediction: 2.0874738343871235 | Target: 0.2845742999916845 | Loss value: 1.8028995343954393

Loss name: MAE
Sample: 4 | Prediction: 1.5995288929771467 | Target: 1.119164165670401 | Loss value: 0.48036472730674573



### `Functions ước lượng hàm số`

In [4]:
from approximate_functions import *

In [4]:
print(f"Approximate function sin: {approx_sin(x=3.14, n=10)}")
print(f"Approximate function cos: {approx_cos(x=3.14, n=10)}")
print(f"Approximate function sinh: {approx_sinh(x=3.14, n=10)}")
print(f"Approximate function cosh: {approx_cosh(x=3.14, n=10)}")

Approximate function sin: 0.001592652393160744
Approximate function cos: -0.9999987352210833
Approximate function sinh: 11.530292029865986
Approximate function cosh: 11.573574824666185


### `Mean difference`

In [5]:
from Mean_difference import *

In [8]:
print(f"Result 1: {mean_difference(yi=100, yhat=99.5, n=2, p=1)}")
print(f"Result 2: {mean_difference(yi=50, yhat=49.5, n=2, p=1)}")
print(f"Result 3: {mean_difference(yi=20, yhat=19.5, n=2, p=1)}")
print(f"Result 4: {mean_difference(yi=0.6, yhat=0.1, n=2, p=1)}")

Result 1: 0.025031328369998107
Result 2: 0.03544417213033135
Result 3: 0.05625552183565574
Result 4: 0.45836890322464546


## II. Trắc nghiệm

`Câu hỏi 1` : Viết function thực hiện đánh giá classification model bằng F1-Score. Function nhận vào 3
giá trị tp, fp, fn và trả về F1-score

In [8]:
def f1_score(tp, fp, fn):
    try:
        if type(tp) != int:
            raise TypeError("tp must be an integer")
        elif type(fp) != int:
            raise TypeError("fp must be an integer")
        elif type(fn) != int:
            raise TypeError("fn must be an integer")

        if tp <= 0 or fp <= 0 or fn <= 0:
            raise ValueError("tp and fp and fn must be greater than 0")

        f1 = 2 * (precision(tp, fp) * recall(tp, fn)) / \
            (precision(tp, fp) + recall(tp, fn))

        return f1
    except Exception as e:
        print(e)

In [9]:
round(f1_score(tp=2, fp=4, fn=5), 2)

0.31

`Answer:` *c) 0.31*

</br>

`Câu hỏi 2` : Viết function is_number nhận input có thể là string hoặc một số **kiểm tra n (một
số) có hợp lệ hay không** (vd: n=’10’, is_number(n) sẽ trả về True ngược lại là False). Đầu ra của
chương trình sau đây là gì?

In [25]:
def is_number(x):
    return True if type(x) == int or type(x) == float else True if x.isdigit() else False

In [27]:
print(is_number(1))
print(is_number('1'))
print(is_number('n'))

True
True
False


`Answer:` *b) True | False*

</br>

`Câu hỏi 3` : Đoạn code dưới đây đang thực hiện activation function nào?

`Answer:` *c) ReLU*

</br>

`Câu hỏi 4` : Viết function thực hiện Sigmoid Function \( f(x) = \frac{1}{1 + e^{-x}} \). Nhận input là x và return kết quả tương ứng trong Sigmoid Function. Đầu ra của chương trình sau đây là gì?

In [29]:
def sigmoid(x):
    return 1 / (1 + math.exp(-x))

In [30]:
round(sigmoid(2), 2)

0.88

`Answer:` *a) 0.88*

</br>

`Câu hỏi 5`:

Viết function thực hiện Elu Function \( f(x) = 
\begin{cases} 
\alpha (e^x - 1) & \text{if } x \leq 0 \\
x & \text{if } x > 0 
\end{cases}
\). Nhận input là x và return kết quả tương ứng trong Elu Function. Đầu ra của chương trình sau đây là gì khi \(\alpha = 0.01\)?

```python

In [31]:
def elu(x, alpha=0.01):
    if x <= 0:
        return alpha * (math.exp(x) - 1)
    else:
        return x

In [46]:
round(elu(-1), 2)

-0.01

`Answer:` *b) -0.01*

</br>

`Câu hỏi 6` : Viết function nhận 2 giá trị x, và tên của activation function act_name **activation
function chỉ có 3 loại (sigmoid, relu, elu)**, thực hiện tính toán activation function tương ứng với
name nhận được trên giá trị của x và trả kết quả. Đầu ra của chương trình sau đây là gì?

In [47]:
def activation_function(x, act_name, alpha=0.01):
    if act_name == "sigmoid":
        return 1 / (1 + math.exp(-x))
    elif act_name == "relu":
        return max(0, x)
    elif act_name == "elu":
        if x <= 0:
            return alpha * (math.exp(x) - 1)
        else:
            return x
    else:
        raise ValueError("Unsupported activation function")

In [49]:
print ( round ( activation_function (x = 3 , act_name ='sigmoid') , 2) )

0.95


`Answer:` *a) 0.95*

</br>

`Câu hỏi 7`:

Viết function tính absolute error \( \text{absolute error} = |y - \hat{y}| \). Nhận input là y và \(\hat{y}\), return về kết quả absolute error tương ứng. Đầu ra của chương trình sau đây là gì?

```python

In [53]:
def absolute_error(y, y_hat):
    return abs(y - y_hat)

In [54]:
print (absolute_error(2 , 9))

7


`Answer:` *a) 7*

</br>

`Câu hỏi 8` : Viết function tính squared error = (y−yˆ)
2
. Nhận input là y và yˆ, return về kết quả squared
error tương ứng. Đầu ra của chương trình sau đây là gì?


In [56]:
def squared_error(y, y_hat):
    return (y - y_hat) ** 2

In [57]:
print (squared_error(2, 1))

1


`Answer:` *a) 1*

</br>

`Câu hỏi 9` : Dựa vào công thức xấp xỉ cos và điều kiện được giới thiệu. Viết function xấp xỉ cos khi nhận x là giá trị muốn tính cos(x) và n là số lần lặp muốn xấp xỉ. Return về kết quả
cos(x) với bậc xấp xỉ tương ứng. Đầu ra của chương trình sau đây là gì?

In [59]:
def approximate_cos(x, n):
    cos_x = 0
    sign = 1
    for i in range(n):
        term = (x ** (2 * i)) / math.factorial(2 * i)
        cos_x += sign * term
        sign *= -1
    return cos_x

In [60]:
print ( round ( approximate_cos ( x =3.14 , n =10) , 2) )

-1.0


`Answer:` *c) -1*

</br>

`Câu hỏi 10` : Dựa vào công thức xấp xỉ sin và điều kiện được giới thiệu. Viết
function xấp xỉ sin khi nhận x là giá trị muốn tính sin(x) và n là số lần lặp muốn xấp xỉ. Return về kết
quả sin(x) với bậc xấp xỉ tương ứng. Đầu ra của chương trình sau đây là gì?

In [62]:
def approximate_sin(x, n):
    sin_x = 0
    sign = 1
    for i in range(n):
        term = (x ** (2 * i + 1)) / math.factorial(2 * i + 1)
        sin_x += sign * term
        sign *= -1
    return sin_x

In [63]:
print ( round ( approximate_sin ( x =3.14 , n =10) , 4) )

0.0016


`Answer:` *a) 0.0016*

</br>

`Câu hỏi 11` : Dựa vào công thức xấp xỉ sinh và điều kiện được giới thiệu các bạn click ở đây. Viết
function xấp xỉ sinh khi nhận x là giá trị muốn tính sinh(x) và n là số lần lặp muốn xấp xỉ. Return về
kết quả sinh(x) với bậc xấp xỉ tương ứng. Đầu ra của chương trình sau đây là gì?

In [66]:
def approximate_sinh(x, n):
    sinh_x = 0
    for i in range(n):
        term = (x ** (2 * i + 1)) / math.factorial(2 * i + 1)
        sinh_x += term
    return sinh_x

In [67]:
print ( round ( approximate_sinh ( x =3.14 , n =10) , 2) )

11.53


`Answer:` *a) 11.53*

</br>

`Câu hỏi 12` : Dựa vào công thức xấp xỉ cosh và điều kiện được giới thiệu. Viết
function xấp xỉ cosh khi nhận x là giá trị muốn tính cosh(x) và n là số lần lặp muốn xấp xỉ. Return về
kết quả cosh(x) với bậc xấp xỉ tương ứng. Đầu ra của chương trình sau đây là gì?

In [69]:
def approximate_cosh(x, n):
    result = 1
    for i in range(1, n + 1):
        result += (x ** (2 * i)) / (factorial(2 * i))
    return result

In [70]:
print ( round ( approximate_cosh ( x =3.14 , n =10) , 2) )

11.57


`Answer:` *a) 11.57*

</br>

`Câu hỏi 13` : Đoạn code nào thực hiện đúng Mean Difference of n
th Root Error được miêu tả ở trên?

`Answer:` *a) (A)*